phoenix-core/src/it/java/org/apache/phoenix/end2end/InvalidateMetadataCacheIT.java

```diff
@@ -0,0 +1,141 @@
+                    " ADD CF.col2 integer");
+            fail("Shouldn't reach heere");
+        } catch (Exception e) {
+            LOGGER.error("Exception while adding column", e);
+            // This is expected
+        }
+    }
+
+    /**
+     * Add FailingPhoenixRegionServerEndpoint as regionserver co-processor.
+     * Make one of the regionserver throw Exception in invalidateServerMetadataCache method.
+     * Make sure that ALTER TABLE ADD COLUMN statement fails.
+     */
+    @Test
+    public void testAddColumnWithOneRSFailing() {
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String dataTableFullName = generateUniqueName();
+        String ddl = getCreateTableStmt(dataTableFullName);
+        HRegionServer regionServerZero = utility.getMiniHBaseCluster().getRegionServer(0);
+        FailingPhoenixRegionServerEndpoint coprocForRS0 =
+                getFailingPhoenixRegionServerEndpoint(regionServerZero);
+        coprocForRS0.throwException();
+        try(Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.createStatement().execute(ddl);
+            conn.createStatement().execute("ALTER TABLE " + dataTableFullName +
+                    " ADD CF.col2 integer");
+            fail("Shouldn't reach here");
+        } catch (Exception e) {
+            LOGGER.error("Exception while adding column", e);
+            // This is expected
+        }
+    }
+
+    /**
+     * Add FailingPhoenixRegionServerEndpoint as regionserver co-processor.
+     * Do not throw Exception or sleep in invalidateServerMetadataCache method for any regionservers
+     * Make sure that ALTER TABLE ADD COLUMN statement succeeds.
+     */
+    @Test
+    public void testAddColumnWithBothRSPassing() {
+        Properties props = PropertiesUtil.deepCopy(TEST_PROPERTIES);
+        String dataTableFullName = generateUniqueName();
+        String ddl = getCreateTableStmt(dataTableFullName);
+        try(Connection conn = DriverManager.getConnection(getUrl(), props)) {
+            conn.createStatement().execute(ddl);
+            conn.createStatement().execute("ALTER TABLE " + dataTableFullName +
+                    " ADD CF.col2 integer");
+        } catch (Throwable t) {
+            fail("Shouldn't reach here");
+        }
+    }
+
+    private String getCreateTableStmt(String tableName) {
+        return   "CREATE TABLE " + tableName +
+                "  (a_string varchar not null, col1 integer" +
+                "  CONSTRAINT pk PRIMARY KEY (a_string)) ";
+    }
+
+    private FailingPhoenixRegionServerEndpoint getFailingPhoenixRegionServerEndpoint(
+            HRegionServer regionServer) {
+        FailingPhoenixRegionServerEndpoint coproc = regionServer
+                .getRegionServerCoprocessorHost()
+                .findCoprocessor(FailingPhoenixRegionServerEndpoint.class);
+        return coproc;
+    }
+}
```

Comment:
You should also add a test where the invalidation is successful on retry

Reply:
Trying to think how can I make it succeed on the retry. Let me try something.

Reply:
Addressed in latest commit. Please review again. Thank you.